includes/embeds/class-amp-youtube-embed-handler.php

```diff
@@ -82,117 +94,311 @@ public function unregister_embed() {
+	 *
+	 * @param Document $dom  Document DOM.
+	 * @param Element  $node YouTube iframe element.
+	 *
+	 * @return Element|false AMP component, otherwise `false`.
+	 */
+	private function get_amp_component( Document $dom, Element $node ) {
+
+		$url        = $node->getAttribute( 'src' );
+		$video_id   = $this->get_video_id_from_url( $url );
+		$attributes = $this->prepare_attributes( $url, $video_id );
+
+		$iframe_props = [ 'title', 'height', 'width' ];
+
+		foreach ( $iframe_props as $iframe_prop ) {
+			if ( ! empty( $node->getAttribute( $iframe_prop ) ) ) {
+				$attributes[ $iframe_prop ] = $node->getAttribute( $iframe_prop );
+			}
+		}
+
+		if ( empty( $attributes['data-videoid'] ) && empty( $attributes['data-live-channelid'] ) ) {
+			$link_node = AMP_DOM_Utils::create_node(
+				$dom,
+				Tag::A,
 				[
-					'href'  => esc_url_raw( $url ),
-					'class' => 'amp-wp-embed-fallback',
-				],
-				esc_html( $url )
+					Attribute::HREF   => esc_url_raw( $url ),
+					Attribute::CLASS_ => 'amp-wp-embed-fallback',
+				]
 			);
+			$link_node->appendChild( $dom->createTextNode( $url ) );
+
+			return $link_node;
+		}
+
+		$amp_node = AMP_DOM_Utils::create_node(
+			$dom,
+			'amp-youtube',
+			$attributes
+		);
+
+		if ( ! empty( $amp_node ) && $amp_node instanceof Element ) {
+			$placeholder = $this->get_placeholder_component( $amp_node, $attributes, $url );
+
+			if ( $placeholder ) {
+				$amp_node->appendChild( $placeholder );
+			}
+		}
+
+		return $amp_node;
+	}
+
+	/**
+	 * Prepare attributes for amp-youtube component.
+	 *
+	 * @param string $url      YouTube video URL.
+	 * @param string $video_id YouTube video ID.
+	 *
+	 * @return array prepared arguments for amp-youtube component.
+	 */
+	private function prepare_attributes( $url, $video_id = '' ) {
+
+		if ( empty( $url ) ) {
+			return [];
+		}
+
+		$attributes = [
+			'layout' => 'responsive',
+			'width'  => $this->args['width'],
+			'height' => $this->args['height'],
+		];
+
+		if ( ! empty( $video_id ) ) {
+			$attributes['data-videoid'] = $video_id;
+		}
+
+		// Find start time of video.
+		$start_time = $this->get_start_time_from_url( $url );
+		if ( ! empty( $start_time ) && 0 < (int) $start_time ) {
+			$attributes['data-param-start'] = (int) $start_time;
+		}
+
+		$query_vars  = [];
+		$query_param = wp_parse_url( $url, PHP_URL_QUERY );
+		wp_parse_str( $query_param, $query_vars );
+		$query_vars = ( is_array( $query_vars ) ) ? $query_vars : [];
+
+		$excluded_param = [ 'start', 'v', 'vi', 'w', 'h' ];
+
+		foreach ( $query_vars as $key => $value ) {
+
+			if ( in_array( $key, $excluded_param, true ) ) {
+				continue;
+			}
+
+			if ( in_array( $key, [ 'autoplay', 'loop' ], true ) ) {
+				$attributes[ $key ] = $value;
+				continue;
+			}
+
+			if ( 'channel' === $key ) {
+				$attributes['data-live-channelid'] = $value;
+				continue;
+			}
+
+			$attributes[ sanitize_key( "data-param-$key" ) ] = esc_attr( $value );
```

Comment:
Escaping shouldn't be done here because the returned attributes array may be sent into the DOM directly, in which case the escaping will be duplicated.

```suggestion
			$attributes[ sanitize_key( "data-param-$key" ) ] = $value;
```